kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt

```diff
@@ -0,0 +1,360 @@
+        expressionTextView.text = expressionToString(expression)
+      }
+    }
+  }
+
+  /**
+   * Run the network trace with the parameters and display the result in an alert dialog.
+   *
+   * @param view of the trace button
+   */
+  fun trace(view: View) {
+    // don't attempt a trace on an unloaded utility network
+    if (utilityNetwork.loadStatus != LoadStatus.LOADED) {
+      return
+    }
+    try {
+      val parameters =
+        UtilityTraceParameters(UtilityTraceType.SUBNETWORK, listOf(startingLocation)).apply {
+          sourceTier?.traceConfiguration?.let {
+            traceConfiguration = it
+          }
+        }
+      val traceFuture = utilityNetwork.traceAsync(parameters)
+      traceFuture.addDoneListener {
+        try {
+          val results = traceFuture.get()
+          (results.firstOrNull() as? UtilityElementTraceResult)?.let { elementResult ->
+            // create an alert dialog
+            AlertDialog.Builder(this).apply {
+              // set the alert dialog title
+              setTitle("Trace result")
+              // show the element result count
+              setMessage(elementResult.elements.count().toString() + " elements found.")
+            }.show()
+          }
+        } catch (e: Exception) {
+          (e.cause?.message + "\nFor a working barrier condition, try \"Transformer Load\" Equal \"15\".").also {
+            Toast.makeText(this, it, Toast.LENGTH_LONG).show()
+            Log.e(TAG, it)
+          }
+        }
+      }
+    } catch (exception: Exception) {
+      ("Error during trace operation: " + exception.message).also {
+        Toast.makeText(this, it, Toast.LENGTH_LONG).show()
+        Log.e(TAG, it)
+      }
+    }
+  }
+
+  /**
+   * Convert the given UtilityTraceConditionalExpression into a string.
+   */
+  private fun expressionToString(expression: UtilityTraceConditionalExpression): String? {
+    when (expression) {
+      // when the expression is a category comparison expression
+      is UtilityCategoryComparison -> {
+        return expression.category.name + " " + expression.comparisonOperator
+      }
+      // when the expression is an attribute comparison expression
+      is UtilityNetworkAttributeComparison -> {
+        (expression.networkAttribute.domain as? CodedValueDomain)?.let { codedValueDomain ->
+          // if there's a coded value domain name
+          val codedValueDomainName = codedValueDomain.codedValues.first {
+            convertToDataType(it.code, expression.networkAttribute.dataType) ==
+                convertToDataType(expression.value, expression.networkAttribute.dataType)
+          }.name
+          return expression.networkAttribute.name + " " + expression.comparisonOperator + " " + codedValueDomainName
+        }
+          ?: return expression.networkAttribute.name + " " + expression.comparisonOperator + " " + (expression.otherNetworkAttribute?.name
+            ?: expression.value)
+      }
```

Comment:
^ This is pretty complex, took me a while to parse out. Not sure if there's a way to separate it out. Or perhaps you could take the first part of the return string out so that the code is a little more compact? ie. 
```
val expressionAttributeAndComparison = expression.networkAttribute.name + " " + expression.comparisonOperator + " "

...

return expressionAttributeAndComparison + codedValueDomainName} ?: expressionAttributeAndComparison + //etc..

```